src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.cs

```diff
@@ -0,0 +1,223 @@
+        /// data is just a blob that is keyed by a string Id.  Data with this ID can be retrieved
+        /// or overwritten.
+        /// 
+        /// The format of the table is:
+        /// 
+        ///  SolutionData
+        ///  --------------------------------------------
+        ///  | Id (primary key, varchar) | Value (blob) |
+        ///  --------------------------------------------
+        /// </summary>
+        private const string SolutionDataTableName = "SolutionData";
+
+        /// <summary>
+        /// Inside the DB we have a table for data that we want associated with a <see cref="Project"/>.
+        /// The data is keyed off of an integral value produced by combining the ID of the Project and
+        /// the ID of the name of the data (see <see cref="SQLitePersistentStorage.ReadStreamAsync(Project, string, CancellationToken)"/>.
+        /// 
+        /// This gives a very efficient integral key, and means that the we only have to store a 
+        /// single mapping from stream name to ID in the string table.
+        /// 
+        /// The format of the table is:
+        /// 
+        ///  ProjectData
+        ///  --------------------------------------------
+        ///  | Id (primary key, integer) | Value (blob) |
+        ///  --------------------------------------------
+        /// </summary>
+        private const string ProjectDataTableName = "ProjectData";
+
+        /// <summary>
+        /// Inside the DB we have a table for data that we want associated with a <see cref="Project"/>.
+        /// The data is keyed off of an integral value produced by combining the ID of the Project and
+        /// the ID of the name of the data (see <see cref="SQLitePersistentStorage.ReadStreamAsync(Project, string, CancellationToken)"/>.
+        /// 
+        /// This gives a very efficient integral key, and means that the we only have to store a 
+        /// single mapping from stream name to ID in the string table.
+        /// 
+        /// The format of the table is:
+        /// 
+        ///  DocumentData
+        ///  --------------------------------------------
+        ///  | Id (primary key, integer) | Value (blob) |
+        ///  --------------------------------------------
+        /// </summary>
+        private const string DocumentDataTableName = "DocumentData";
+
+        private const string IdColumnName = "Id";
+        private const string ValueColumnName = "Value";
+        private const string DataColumnName = "Data";
+
+        static SQLitePersistentStorage()
+        {
+            // Necessary for SQLitePCL to call into SQLitePCL.
+            SQLitePCL.Batteries_V2.Init();
+        }
+
+        private readonly CancellationTokenSource _shutdownTokenSource = new CancellationTokenSource();
+
+        // Accessors that allow us to retrieve/store data into specific DB tables.  The
+        // core Accessor type has logic that we to share across all reading/writing, while
```

Comment:
This comment is about the implementation of these types, and isn't really helpful in this context.